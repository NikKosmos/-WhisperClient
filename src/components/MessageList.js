import React from 'react'
import PropTypes from 'prop-types'

class MessageList extends React.Component {
    
    renderMessageList = () => {
        const {messages} = this.props;

        if (messages.length) {
            return messages.map(function (item) {
                return (
                    <p>{item.messages}</p>
                )
            })
        }
    }

    render() {
        return (
            <div className="MessageList">
                {this.renderMessageList()}
            </div>
        )
    }
}

MessageList.propTypes = {
    messages: PropTypes.arrayOf(PropTypes.string)
}

export {MessageList}