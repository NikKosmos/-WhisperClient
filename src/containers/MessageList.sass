.messageList
    overflow-y: scroll
    height: 380px