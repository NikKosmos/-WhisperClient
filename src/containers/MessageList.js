import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { fetchMessagesList } from '../actions/messageListActions'
import { fetchNewMessages } from '../actions/messageListActions'
import { MessageFrame } from '../components/MessageFrame'
import { NewMessageInput } from '../components/NewMessageInput'
import { updateInterval } from '../applicationSettings'
import './MessageList.sass'

class MessageList extends React.Component {

    constructor(props) {
        super(props)
        
        this.messageListRef = React.createRef()

        this.state = {
            enableScrollDown: true,
            previousMessagesLength: 0,
        }
    }

    componentDidMount = () => {
        this.fetchMessages()
        this.initializeMessagesListUpdateTimer()
    }

    componentWillUnmount() {
        clearInterval(this.timerID)
      }

    componentDidUpdate = () => {
        const { messages } = this.props;
        const scrollDownShift = 30
        
        if (!messages) return
        
        const messagesLength = messages.length
        const previousMessagesLength = this.state.previousMessagesLength
        
        if (messagesLength === previousMessagesLength) return

        this.setState({
            previousMessagesLength: messagesLength,
        })

        const { current } = this.messageListRef

        current.scrollTop += scrollDownShift

        this.scrollDownIfEnabled()
    }

    initializeMessagesListUpdateTimer = () => {
        this.timerID = setInterval(() => this.fetchNewMessages(), updateInterval)
    }
    
    fetchMessages = () => {
        const currentChat = this.props.currentChat
        const messages = this.props.messages

        if(!currentChat || !currentChat._id) return
        if(!messages) return

        const time = (new Date()).getTime();

        const oldestMessageTime = messages.length ? messages[0].time : time

        this.props.fetchMessagesList(currentChat._id, oldestMessageTime)
    }

    fetchNewMessages = () => {
        const currentChat = this.props.currentChat
        const messages = this.props.messages

        if(!currentChat || !currentChat._id) return
        if(!messages) return

        const lastMessageIndex = messages.length - 1

        if (lastMessageIndex < 0) {
            this.fetchMessages()
            return
        }

        this.props.fetchNewMessages(currentChat._id, messages[lastMessageIndex].time)
    }

    scrollDownIfEnabled = () => {
        if (this.state.enableScrollDown) {

            this.setState({
                enableScrollDown: false,
            })

            this.scrollDown()
        }
    }

    scrollDown = () => {
        const { current } = this.messageListRef

        current.scrollTop = current.scrollHeight
    }

    renderMessageList = () => {
        const { messages } = this.props;
        const messagesLength = messages.length

        if (messages && messagesLength) {
            return messages.map(function (item, index) {
                return (
                    <MessageFrame key = { index } message = { item } />
                    )
                })
        }
    }

    onScrollDownClick = () => {
        this.scrollDown()
    }

    onScroll = () => {
        const minScrollTop = 30
        const { current } = this.messageListRef

        if (current.scrollTop < minScrollTop) {
            this.fetchMessages()
        }
    }

    sendNewMessage = (newMessage) => {
        this.props.sendNewMessage(newMessage)
        
        this.setState({
            enableScrollDown: true,
        })
    }

    render() {
        return (
            <div>
                <button onClick = { this.onScrollDownClick}>
                    Scroll down
                </button>
                <div ref = { this.messageListRef } className="messageList" onScroll = { this.onScroll }>
                    {this.renderMessageList()}
                </div>
                <NewMessageInput
                    onSubmitNewMessage = { this.sendNewMessage }
                />
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        messages: state.messages,
        currentChat: state.currentChat,
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        fetchMessagesList: (chatId, oldestMessageTime) => dispatch(fetchMessagesList(chatId, oldestMessageTime)),
        fetchNewMessages: (chatId, newestMessageTime) => dispatch(fetchNewMessages(chatId, newestMessageTime)),
    }
}

MessageList.propTypes = {
    messages: PropTypes.arrayOf(PropTypes.shape({
        message: PropTypes.exact({
            chatId: PropTypes.number.isRequired,
            authorName: PropTypes.string.isRequired,
            authorEmail: PropTypes.string.isRequired,
            time: PropTypes.oneOfType([PropTypes.number, PropTypes.string, PropTypes.instanceOf(Date)]).isRequired,
            text: PropTypes.string.isRequired,
            wasMessageReceived: PropTypes.bool,
        })
    }))
}

export { MessageList }
export default connect(mapStateToProps, mapDispatchToProps)(MessageList);