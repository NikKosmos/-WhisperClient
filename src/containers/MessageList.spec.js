import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import configureStore from 'redux-mock-store'

import { MessageList } from './MessageList'
import ConnectedMessageList from './MessageList'

describe('MessageList', () => {
    const mockMessages = [{
        _id: '0',
        chatId: '0',
        authorName: 'captain Nemo',
        authorEmail: 'nemo@mail.com',
        time: '2019-12-25T16:40:12.526Z',
        text: 'some message',
        wasMessageReceived: true,
    }]
    const mockFetchMessagesList = jest.fn()
    let testedComponent

    beforeEach(() => {
        testedComponent = shallow(<MessageList messages={ mockMessages } fetchMessagesList = { mockFetchMessagesList }/>)
    })

    it('should render correctly', () => {

        expect(shallowToJson(testedComponent)).toMatchSnapshot();
    })

})
