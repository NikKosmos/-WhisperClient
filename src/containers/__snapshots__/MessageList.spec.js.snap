// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageList should render correctly 1`] = `
<Fragment>
  <button
    onClick={[Function]}
  >
    Scroll down
  </button>
  <div
    className="messageList"
    onScroll={[Function]}
  >
    <MessageFrame
      key="0"
      message={
        Object {
          "_id": "0",
          "authorEmail": "nemo@mail.com",
          "authorName": "captain Nemo",
          "chatId": "0",
          "text": "some message",
          "time": "2019-12-25T16:40:12.526Z",
          "wasMessageReceived": true,
        }
      }
    />
  </div>
  <NewMessageInput
    onSubmitNewMessage={[Function]}
  />
</Fragment>
`;
